File: ml-benchmark/autolifter/dac/mtp.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = -4;
Config SampleIntMax = 4;
Config SampleSize = 8;
Config ComposeNum = 3;
Config NonLinear = true;

Inductive List = cons {Int, List} | nil Unit;

length = fix (\f: List->Int. \x: List.
  match x with
    cons {h, t} -> + (f t) 1
  | nil _ -> 0
  end
);

concat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

split = \xs: List. fix (\f: List->Int->{List, List}. \x: List. \n: Int.
  if (< n 1) then {nil unit, x}
  else match x with
    cons {h, t} -> let res = f t (- n 1) in {cons {h, res.1}, res.2}
  | _ -> {nil unit, nil unit}
  end
) xs (/ (length xs) 2);

dac = \v: List -> Int. \xs: List.
  let run = fix (\f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {_, nil _} -> xs
    | _ -> let sp = split xs in
      concat (f sp.1) (f sp.2)
    end
  ) in v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mtp = \xs: List. (fix (lambda f: List -> {Int, Int}. lambda l: List.
  match l with
    nil _ -> {1, 1}
  | cons {h, t} ->
    let res = f t in
    let tot = * h res.2 in
    {max res.1 tot, tot}
  end
) xs).1;

main = dac mtp;

File: ml-benchmark/autolifter/dac/mtp.ml
type lst = Cons of int * lst | Nil of unit

let rec length x =
  match x with
  | Cons (_, t) -> length t + 1
  | Nil _ -> 0

let rec concat x y =
  match x with
  | Cons (h, t) -> Cons (h, concat t y)
  | Nil _ -> y

let split xs =
  let rec f x n =
    if n < 1 then (Nil (), x)
    else
      match x with
      | Cons (h, t) ->
          let res = f t (n - 1) in
          match res with
          | (p1, p2) -> (Cons (h, p1), p2)
      | Nil _ -> (Nil (), Nil ())
  in
  f xs (length xs / 2)

let dac v xs =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (_, t) ->
        match t with
        | Nil _ -> xs
        | _ ->
            let sp = split xs in
            match sp with
            | (a, b) -> concat (run a) (run b)
  in
  v (run xs)

let max a b =
  if a < b then b else a

let mtp xs =
  let rec f l =
    match l with
    | Nil _ -> (1, 1)
    | Cons (h, t) ->
        let res = f t in
        match res with
        | (r1, r2) ->
            let tot = h * r2 in
            let m = max r1 tot in
            (m, tot)
  in
  match f xs with
  | (a, b) -> a

let program = dac mtp